### PR TITLE
Location property

### PR DESCRIPTION
**Pull request by ericshermancs**

### CODE REVIEW

**Comment by Huddie** [comment]
There are merge conflicts that need to be fixed.

File: Dangerfile [hunk 1]
```diff
@@ -12,15 +12,19 @@ warn("Big PR") if git.lines_of_code > 500
 fail("fdescribe left in tests") if `grep -r fdescribe specs/ `.length > 1
 fail("fit left in tests") if `grep -r fit specs/ `.length > 1
 
-forgot_tests = !git.modified_files.include?("/cunyfirstapi/tests/tests.py")
+forgot_tests = !git.modified_files.include?("/tests/run_tests.py")
```
Comment (Huddie) [inline Dangerfile:15]: Keep this version

File: Dangerfile [hunk 2]
```diff
@@ -12,15 +12,19 @@ warn("Big PR") if git.lines_of_code > 500
...
 
 if forgot_tests and not declared_trivial
-	warn("It appears that you forgot to add a Unit Test to the test file.\n Please add a test and upload the new version.\n The test file can currently be found at: cunyfirstapi/tests/tests.py")
+	warn("It appears that you forgot to add a Unit Test to the test file.\n Please add a test and upload the new version.\n The test file can currently be found at: tests/run_tests.py")
```
Comment (Huddie) [inline Dangerfile:18]: Keep this version

File: Depfiles/dependencies.comp [hunk 3]
```diff
@@ -0,0 +1 @@
+vlucas/phpdotenv
```
Comment (Huddie) [inline Depfiles/dependencies.comp]: Remove this file, its unnecessary

File: Depfiles/depinstall.sh [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
```
Comment (Huddie) [inline Depfiles/depinstall.sh:1]: This file was already fixed, don't add it.

File: Depfiles/dependencies.pip [hunk 5]
```diff
@@ -0,0 +1,3 @@
+requests
```
Comment (Huddie) [inline Depfiles/dependencies.pip]: This file already exists, done add it

File: cunyfirstapi/__init__.py [hunk 6]
```diff
@@ -1,8 +1,10 @@
 name = "cunyfirstapi"
 from .cunyfirstapi import *
-from . import login as cuny_login
+from . import login 
 from . import constants
+#from .tests import run_tests
```
Comment (Huddie) [inline cunyfirstapi/__init__.py]: This file was already changed, dont add it.

File: tests/run_tests.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
+###***********************************###
```
Comment (Huddie) [inline tests/run_tests.py]: This file already exists, add test cases, dont change anyhting else

File: tests/tests.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
+###***********************************###
```
Comment (Huddie) [inline tests/tests.py]: This file should not exist.

**Review by Huddie — CHANGES_REQUESTED** [review]

File: Depfiles/dependencies.comp (hunk 3 above)
Comment (Huddie) [reply Depfiles/dependencies.comp]: ^^ Remove this fine

File: cunyfirstapi/cunyfirstapi.py [hunk 9]
```diff
@@ -15,13 +15,13 @@
...
-from . import login as cuny_login
-from . import constants
-from .student_center import Student_Center, Student_Center_Action
-from .grades import Student_Grades, Student_Grades_Action
-from .actions_locations import ActionObject, Location
-from .transcript import Transcript_Page, Transcript_Page_Action
-from .locations_enum import Locations
+from cunyfirstapi import login as cuny_login
+from cunyfirstapi import constants
+from cunyfirstapi.student_center import Student_Center, Student_Center_Action
```
[the remark this answers is not in the document]
Comment (Huddie) [reply cunyfirstapi/cunyfirstapi.py]: Hound ^^^

File: cunyfirstapi/cunyfirstapi.py [hunk 10]
```diff
@@ -15,13 +15,13 @@
...
+from cunyfirstapi.grades import Student_Grades, Student_Grades_Action
```
[the remark this answers is not in the document]
Comment (Huddie) [reply cunyfirstapi/cunyfirstapi.py]: Hound ^^

File: cunyfirstapi/cunyfirstapi.py [hunk 11]
```diff
@@ -15,13 +15,13 @@
...
+from cunyfirstapi.actions_locations import ActionObject, Location
```
[the remark this answers is not in the document]
Comment (Huddie) [reply cunyfirstapi/cunyfirstapi.py]: Hound ^^

File: cunyfirstapi/cunyfirstapi.py [hunk 12]
```diff
@@ -15,13 +15,13 @@
...
+from cunyfirstapi.transcript import Transcript_Page, Transcript_Page_Action
```
[the remark this answers is not in the document]
Comment (Huddie) [reply cunyfirstapi/cunyfirstapi.py]: Hound ^^

File: cunyfirstapi/transcript.py [hunk 13]
```diff
@@ -12,9 +12,9 @@
 from lxml import html, etree
 from os.path import join
 import re
-from . import constants
-from .helper import get_semester
-from .actions_locations import ActionObject, Location
+from cunyfirstapi import constants
+from cunyfirstapi.helper import get_semester
```
[the remark this answers is not in the document]
Comment (Huddie) [reply cunyfirstapi/transcript.py:16]: Hound ^^

File: tests/redacted_stdout.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
+import io, sys, re
```
[the remark this answers is not in the document]
Comment (Huddie) [reply tests/redacted_stdout.py:1]: Hound ^^

File: tests/redacted_stdout.py [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+
+class RedactedPrint:
+
+    def __init__(self, option, redacted_list):
+        self.origOut = None
+        self.option = option
+        self.redacted_list = redacted_list
+
+    def enable(self):
+
```
Comment (Huddie) [inline tests/redacted_stdout.py:15]: remove newline

File: cunyfirstapi/class_search.py [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
+            'Accept-Encoding': 'gzip, deflate, br',
+            'Accept-Language': 'en-US,en;q=0.9',
+            'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3396.99 Safari/537.36',
+            'Content-Type': 'application/x-www-form-urlencoded',
+            'Accept': '*/*',
+            'Referer': constants.CUNY_FIRST_CLASS_SEARCH_URL,
+            'Connection': 'keep-alive',
+        }  
+        '''  
+        ICAction: Tell the server we are gonna search for a class now
```
Comment (Huddie) [inline cunyfirstapi/class_search.py]: This comment applies to all the docstrings below. These are written well! But. Any def of a variable (term, course career, institution ...etc) need to be placed in 1 docstring above the function declaration.

Comment (Huddie) [reply cunyfirstapi/class_search.py]: Short 1/2 sentence blurb should go above these broken up dictionary form pieces explaining which part they are.

File: cunyfirstapi/cunyfirstapi.py [hunk 17]
```diff
@@ -15,13 +15,13 @@
...
 from os.path import join, dirname
-from . import login as cuny_login
-from . import constants
-from .student_center import Student_Center, Student_Center_Action
-from .grades import Student_Grades, Student_Grades_Action
-from .actions_locations import ActionObject, Location
-from .transcript import Transcript_Page, Transcript_Page_Action
-from .locations_enum import Locations
+from cunyfirstapi import login as cuny_login
+from cunyfirstapi import constants
```
[the remark this answers is not in the document]
Comment (Huddie) [reply cunyfirstapi/cunyfirstapi.py]: Hound ^^

File: cunyfirstapi/cunyfirstapi.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Huddie) [reply cunyfirstapi/cunyfirstapi.py]: ^^

File: cunyfirstapi/class_search.py [hunk 18]
```diff
@@ -0,0 +1,304 @@
...
+                        'E' -> is exactly
+    instructor_last_name: part or whole of instructor's last name, match by above
+    '''   
+    def submit_search(self, institution, term, course_number, 
+                        subject='', course_number_match='E', 
+                        course_career='', course_attribute='', 
+                        course_attribute_value='', 
+                        requirement_designation='',
+                        open_classes_only=True, session='', 
+                        mode_of_instruction='',
```
[the remark this answers is not in the document]
Comment (Huddie) [reply cunyfirstapi/class_search.py]: Session i son wrong line here ^^

File: cunyfirstapi/class_search.py [hunk 19]
```diff
@@ -0,0 +1,304 @@
...
+                        meeting_start_time='', 
+                        meeting_end_time_match='LE',
+                        meeting_end_time='', days_of_week_match='', 
+                        days_of_week=None, class_number='', 
+                        course_keyword='',
+                        minimum_units_match='GE', minimum_units='', 
+                        maximum_units_match='LE', maximum_units='', 
+                        course_component='', campus='', _location='', 
+                        instructor_last_name_match='B', 
+                        instructor_last_name='', parsed=True):
```
[the remark this answers is not in the document]
Comment (Huddie) [reply cunyfirstapi/class_search.py]: Put each variable on its own line. This is difficult to read

**Comment by Huddie** [comment]
Okay follow up requirements

1. Hound caught a few unused imports
2. Put each variable on its own line (See my comments)
3. We need another 4/5/6 test cases on this.

Then we should be good for another full review and hopefully merge

File: tests/run_tests.py [hunk 20]
```diff
@@ -39,6 +39,46 @@ def test(self):
             api.login()
             self.assertTrue(api.is_logged_in())
 
+class TestClassSearch(unittest.TestCase):
```
Comment (Huddie) [inline tests/run_tests.py:42]: You sure your testing for every edge case you can think of? The point of these tests is to catch a mistake later on if someone changes the code.

Also make sure to test for things that should fail, a CS class doesnt show up when searching etc. **A closed class doesnt show when searching open.** <--- make sure that test exists. BOTH WAYS..

Finally, its important that the tesst are all separate.  Have a single assertTrue( means that if it fails you have 0 idea which of the tests failed.

**Review by Huddie — CHANGES_REQUESTED** [review]
Gotta fix + add a few tests

**Review by Huddie — APPROVED** [review]
thumbs up. 

We can deal with the dangerfile later.
